File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {number}
 */
function getMiddle(head){
 
    let slow = head;
    let fast = head;
    while(fast && fast.next && fast.next.next){
        slow = slow.next;
        fast = fast.next.next
    }
    return slow
} 
function reverse(head){
      let curr = head;
    let prev = null;

    while (curr !== null) {
        let temp = curr.next;
        curr.next = prev;
        prev = curr;
        curr = temp;
    }
    return prev
}
var pairSum = function(head) {
    let middle = getMiddle(head);
    let temp = middle.next;
    middle.next = null;
    let newhead = reverse(temp);
    let maxsum=0
 
    while(head!=null){
        maxsum = Math.max(maxsum , head.val + newhead.val);
        head = head.next;
        newhead = newhead.next


    }

    return maxsum
};

File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* getMiddle(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        ListNode* prev = NULL;
        while (fast && fast->next) {
            prev = slow;
            slow = slow->next;
            fast = fast->next->next;
        }
        return prev;
    }
    ListNode* reverse(ListNode* head) {
        ListNode* prev = NULL;
        ListNode* current = head;
        while (current) {
            ListNode* temp = current->next;
            current->next = prev;
            prev = current;
            current = temp;
        }

        return prev;
    }
    int pairSum(ListNode* head) {
        int maxsum = INT_MIN;
        ListNode* mid = getMiddle(head);
        ListNode* midnext = mid->next;
        mid->next = NULL;
        ListNode* reversed = reverse(midnext);

        while (head && reversed) {
            int sum = head->val + reversed->val;
            if (sum > maxsum) {
                maxsum = sum;
            }
            head = head->next;
            reversed = reversed->next;
        }

        return maxsum;
    }
};